### ISSUE TITLE
Excessive CPU usage

### ISSUE THREAD

**Issue by tones111** [ArturKovacs/emulsion#13]
When running on X I see X using about 70% of a core and emulsion using about 60%, even without an image loaded.

emulsion CPU looks to increase as it's loading images but eventually settles back at the 60% level.  It would be nice if it used less resources while idling.

**Comment by tuncer** [comment]
Can you try https://github.com/ArturKovacs/emulsion/pull/17? Since gelatin is fetched from crates.io by default, you'll have to modify the dependency to be local: `gelatin = { path = "subcrates/gelatin" }`

**Comment by tuncer** [comment]
@ArturKovacs, what do you think about depending on the in-tree gelatin by default?

**Comment by ArturKovacs** [comment]
@tuncer as a matter of fact I didn't want gelatin published but when bulishing any crate all of its dependencies must also be published which forced this clumsy setup. But maybe what you are saying is that all commits to the main should just have it set to path and then only set it to a version when publishing. In which case that's what we should do from now if I think about it.

I'll check out the PR.

**Comment by ArturKovacs** [comment]
I'm closing this as the fix was merged.

**Comment by nivkner** [comment]
I don't think the fix in #17 is enough by itself. what it does is move to an idle state when exactly nothing happens.

while emulsion is running, without any image loaded and after emulsion moved to the idle state , if you move the mouse, without dragging anything or holding space, the CPU percentages jump back up, without the mouse even being close to the window.

ideally emulsion would only do a lot of work after detecting that something has to actually change on screen, and otherwise eagerly move to the idle state.

**Comment by ArturKovacs** [comment]
@nivkner Thanks for pointing this out. I'm reopening this and I will see what can be done about that. But if you'd like to look for a solution yourself don't hesitate to do so.

**Comment by tuncer** [comment]
Is it possible that `redraw_needed` and `request_redraw` are not damage aware?

**Comment by tuncer** [comment]
@nivkner, how much load do you observe? To me it looks like "normal" redraw load caused by cursor movement triggering damage events. The difference on my machine between emulsion on X11 and a glutin hello world calling `swap_buffers_with_damage()` on Wayland, while moving the cursor around, was less than 2%.

Do you know how to optimize it further?

**Comment by tuncer** [comment]
@ArturKovacs, does it cause similar load on Windows?

**Comment by nivkner** [comment]
I am using an intel i5-4460  CPU,
while emulsion is active, moving the mouse causes a CPU load of around 30-50% per core
after waiting for a while the load drops to less than 10% per core, close to the state without emulsion.

I tried comparing the to `damage` example in the glutin repository:
on X11 `swap_buffers_with_damage_supported` returns `false`
because damage is only supported when `X11Context` is EGL, not GLX.

instead I tried the example on wayland, where I got less than 10% per-core,
but I couldn't get emulsion to run on wayland,
kept getting "application not responding" shortly after starting it.

**Comment by nivkner** [comment]
as for how to optimize, a good place to start might be this
https://github.com/ArturKovacs/emulsion/blob/3d1b6537ea2fff52496fdb1e6a38be782edf2165/subcrates/gelatin/src/window.rs#L322-L326

**Comment by tuncer** [comment]
Yes, it hangs in set_title() on Wayland right now. See #11.

I get 5..7% by just moving the cursor around, and like I said, the difference between emulsion on X11 and `swap_buffers_with_damage` on Wayland was less than 2%. That's why I didn't focus on `redraw_needed()` and instead wondered if there's room to optimize at all. That said, I agree that `redraw_needed()` should either be completed or dropped.

I suppose there's a way to enable redraw rect visualization with glium, too.

**Comment by nivkner** [comment]
what was your CPU load before #17 ? if it was 5-7% before, than its always going to look like there is no room for optimization.

I tried comparing emulsion 1.9 with master, on both my desktop and laptop, both using X11:

on the desktop with i5-4460:
both 1.9 and master have the behavior of 30-50% when moving the mouse,
and less than 10% after waiting.

on the laptop with i3-8130U:
neither 1.9 nor master drop the CPU load when not doing anything, at least not consistently.
load stays at 50-70% per core, except for one core which stays at less than 10%

this discrepancy might have something to do with OS scheduling, but it does suggest that #17 didn't actually change anything in my case.

while not that useful on its own, changing `redraw_needed` to always be false, makes the CPU load stay low on both computers regardless of other interactions, so if the time spent with `redraw_needed() == false` could be maximized, it would definitively make a difference.

**Comment by tuncer** [comment]
On X11, with or without a compositor, when idling, emulsion sits at 0% for me. Both in master and #34. On Wayland, with #34, it's eating cpu during idle, but the set_title() bug has been resolved and it's responsive.

To compare, I observed cursor damage load of a relm (gtk-rs) gui and a simple GTK demo app. Both cause as much load on X11+compositor and Wayland as emulsion does, plus higher load on X11 without compositor. But that's to be expected since GTK3 relies heavily on the existence of a compositor.

Meanwhile, FOX toolkit's calculator and a Tcl/Tk app (with theming) don't go above 4% during cursor movement damage.

Let's figure out what's causing load while idling on your machine. To that end, I'd suggest we come up with a redraw/damage debug method.

**Comment by tuncer** [comment]
On X11, with or without a compositor, when idling, emulsion sits at 0% for me. Both in master and #34. On Wayland, with #34, it's eating cpu during idle, but the set_title() bug has been resolved and it's responsive.

To compare, I observed cursor damage load of a relm (gtk-rs) gui and a simple GTK demo app. Both cause as much load on X11+compositor and Wayland as emulsion does, plus higher load on X11 without compositor. But that's to be expected since GTK3 relies heavily on the existence of a compositor.

Meanwhile, FOX toolkit's calculator and a Tcl/Tk app (with theming) don't go above 4% during cursor movement damage.

To be clear, I didn't measure gpu load. Plus, comparing the rendering architecture of FOX and Tk vs GTK or Emulsion is, obviously, not fair. So, take that into account. The point of comparing them is to see if and how much unnecessary load emulsion creates relative to other apps.

Let's figure out what's causing load while idling on your machine. To that end, I'd suggest we come up with a redraw/damage debug method.
